gems/pending/MiqVm/MiqRhevmVm.rb

```diff
@@ -4,16 +4,158 @@ class MiqRhevmVm < MiqVm
+    when "nfs"
+      add_nfs_mount(storage_obj)
+      nfs_file_path(storage_obj, disk)
+    else
+      lun_file_path(storage_obj, disk)
+    end
+  end
+
+  def nfs_mount_root
+    @nfs_mount_root ||= @ost.nfs_mount_root || "/mnt/#{@rhevmVm.attributes[:id]}"
+  end
+
+  def nfs_file_path(storage_obj, disk)
+    storage_id  = storage_obj.attributes[:id]
+    disk_id     = disk.attributes[:id]
+    image_id    = disk.attributes[:image_id]
+    mount_point = nfs_mounts[storage_id][:mount_point]
+
+    ::File.join(mount_point, storage_id, 'images', disk_id, image_id)
+  end
+
+  def lun_file_path(storage_obj, disk)
+    storage_id = storage_obj.attributes[:id]
+    disk_id    = disk[:image_id].blank? ? disk.attributes[:id] : disk.attributes[:image_id]
+
+    ::File.join('/dev', storage_id, disk_id)
+  end
+
+  def storage_domains_by_id
+    @storage_domains_by_id ||= @rhevm.storagedomains.each_with_object({}) { |sd, sdh| sdh[sd.attributes[:id]] = sd }
+  end
+
+  #
+  # Returns uri and mount points, hashed by storage ID.
+  #
+  def add_nfs_mount(storage_obj)
+    storage_id = storage_obj.attributes[:id]
+    return if nfs_mounts[storage_id]
+
+    mount_point = ::File.join(nfs_mount_root, nfs_mount_dir(storage_obj))
+    nfs_mounts[storage_id] = {:uri => "nfs://#{nfs_uri(storage_obj)}", :mount_point => mount_point}
+  end
+
+  def nfs_uri(storage_obj)
+    storage = storage_obj.attributes[:storage]
+    "#{storage[:address]}:#{storage[:path]}"
+  end
+
+  def nfs_mounts
+    @nfs_mounts ||= {}
+  end
+
+  def nfs_mount_dir(storage_obj)
+    nfs_uri(storage_obj).gsub("_", "__").gsub("/", "_")
+  end
+
+  def mount_storage
+    require 'util/mount/miq_nfs_session'
+    log_header = "MIQ(MiqRhevmVm.mount_storage)"
+    $log.info "#{log_header} called"
+
+    @ost.nfs_storage_mounted = false
+
+    if nfs_mounts.empty?
+      $log.info "#{log_header} storage mount not needed."
+      return
+    end
+
+    begin
+      FileUtils.mkdir_p(nfs_mount_root) unless File.directory?(nfs_mount_root)
+      nfs_mounts.each do |storage_id, mount_info|
+        $log.info "#{log_header} Mounting #{mount_info[:uri]} on #{mount_info[:mount_point]} for #{storage_id}"
+        MiqNfsSession.new(mount_info).connect
+        @ost.nfs_storage_mounted = true
+      end
+      $log.info "#{log_header} - mount:\n#{`mount`}"
+    rescue
+      $log.error "#{log_header} Unable to mount all items from <#{nfs_mount_root}>"
+      unmount_storage
+      raise $!
+    end
+  end
+
+  # Moved from MIQExtract.rb
+  def unmount_storage
+    log_header = "MIQ(MiqRhevmVm.unmount_storage)"
+    return if nfs_mounts.empty?
+    begin
+      $log.warn "#{log_header} Unmount all items from <#{nfs_mount_root}>"
+      nfs_mounts.each_value { |mnt| MiqNfsSession.disconnect(mnt[:mount_point]) }
+      FileUtils.rm_rf(nfs_mount_root)
+      @ost.nfs_storage_mounted = false
+    rescue
+      $log.warn "#{log_header} Failed to unmount all items from <#{nfs_mount_root}>.  Reason: <#{$!}>"
```

Comment:
I know you like `$!`, but can we rescue a named exception here and log the exception.message to make it more explicit?